### PR TITLE
HPCC-32144 Investigate using std::shared_mutex for read-write locks

### PR DESCRIPTION
**Pull request by richardkchapman**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: testing/unittests/jlibtests.cpp [hunk 1]
```diff
@@ -3247,7 +3323,7 @@ class AtomicTimingStressTest : public CppUnit::TestFixture
     class Null
     {};
 
-    const unsigned numIterations = 1000000;
+    const unsigned numIterations = 1000;
```
Comment (ghalliday) [inline testing/unittests/jlibtests.cpp:3326]: Should change back - or possibly reduce by a factor of 10 rather than 1000

**Review by ghalliday — COMMENTED** [review]
@richardkchapman one trivial comment

File: testing/unittests/jlibtests.cpp (hunk 1 above)
Comment (richardkchapman) [reply testing/unittests/jlibtests.cpp]: Done, and force-pushed

**Comment by ghalliday** [comment]
Running on my machine I get the following:

Existing implementation:
12:31:54.037650 1559108 UNK UNK 10:1@1/1 threads(10) 209ns/iteration lost(208193)
12:31:57.647563 1559108 UNK UNK 20:1@1/1 threads(20) 180ns/iteration lost(142393)
12:31:59.612704 1559108 UNK UNK 10:10@1/1 threads(10) 196ns/iteration lost(125634)

New implementation:
12:31:05.298590 1558493 UNK UNK 10:1@1/1 threads(10) 155ns/iteration lost(242681)
12:31:07.699419 1558493 UNK UNK 20:1@1/1 threads(20) 119ns/iteration lost(333825)
12:31:08.953719 1558493 UNK UNK 10:10@1/1 threads(10) 125ns/iteration lost(224516)

**Comment by ghalliday** [comment]
Note to self: Merge this code, but rerun the tests above again with restricted cpu affinity to see how that affects performance, and include those numbers in the jira.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**